File: ddns/provider/cloudns.py
# coding=utf-8
"""
ClouDNS API
@doc: https://www.cloudns.net/wiki/
@author: NewFuture
"""

from ._base import BaseProvider, TYPE_FORM


class CloudnsProvider(BaseProvider):
    """
    ClouDNS API
    ClouDNS 接口解析操作库
    """

    endpoint = "https://api.cloudns.net"
    content_type = TYPE_FORM
    DEFAULT_TTL = 60  # Minimum TTL (60 seconds) for faster updates

    def _request(self, action, **params):
        # type: (str, **(str | int | None)) -> dict | None
        """
        发送请求数据，自动添加认证参数

        Send request to ClouDNS API with authentication.

        Args:
            action (str): API endpoint path
            params: API parameters

        Returns:
            dict|None: Response data or None on error
        """
        # Add authentication
        params["auth-id"] = self.id
        params["auth-password"] = self.token

        # Filter None values
        params = {k: v for k, v in params.items() if v is not None}

        data = self._http("POST", action, body=params)

        # ClouDNS API returns different formats:
        # - Success: {"status": "Success", ...}
        # - Failure: {"status": "Failed", "statusDescription": "..."}
        # - Query: raw data (e.g., {"id1": {...}, "id2": {...}}) without status field

        # Check for explicit error status
        if data and isinstance(data, dict) and data.get("status") == "Failed":
            error_msg = data.get("statusDescription", "Unknown error")
            self.logger.warning("ClouDNS API error: %s", error_msg)
            return None

        return data

    def _query_zone_id(self, domain):
        # type: (str) -> str | None
        """
        查询域名区域ID

        Query zone ID for domain.
        ClouDNS uses domain name directly as zone identifier.

        Args:
            domain (str): Domain name

        Returns:
            str: Domain name (used as zone ID)
        """
        # ClouDNS uses domain-name directly, no numeric zone ID
        return domain

    def _query_record(self, zone_id, subdomain, main_domain, record_type, line, extra):
        # type: (str, str, str, str, str | None, dict | None) -> dict | None
        """
        查询DNS记录

        Query DNS record.
        https://www.cloudns.net/wiki/article/57/

        Args:
            zone_id (str): Zone ID (domain name for ClouDNS)
            subdomain (str): Subdomain
            main_domain (str): Main domain
            record_type (str): Record type (A, AAAA, etc.)
            line (str|None): Line (not used by ClouDNS)
            extra (dict|None): Extra parameters

        Returns:
            dict|None: Record data or None if not found
        """
        # For @ (root) records, ClouDNS uses empty string
        host = "" if subdomain == "@" else subdomain

        params = {"domain-name": zone_id, "host": host, "type": record_type}
        data = self._request("/dns/records.json", **params)

        if not data or not isinstance(data, dict):
            return None

        # Check if it's an error response (has status field)
        if "status" in data:
            return None

        # ClouDNS returns records as dict: {"id1": {record_data}, "id2": {record_data}}
        # Find matching record
        for record in data.values():
            record_host = record.get("host", "")
            # Match host (handle both "" and "@" for root records)
            if record_host == host or (subdomain == "@" and record_host in ("", "@")):
                if record.get("type") == record_type:
                    return record

        return None

    def _create_record(self, zone_id, subdomain, main_domain, value, record_type, ttl, line, extra):
        # type: (str, str, str, str, str, int | None, str | None, dict | None) -> bool
        """
        创建DNS记录

        Create new DNS record.
        https://www.cloudns.net/wiki/article/58/

        Args:
            zone_id (str): Zone ID (domain name)
            subdomain (str): Subdomain
            main_domain (str): Main domain
            value (str): Record value (IP address)
            record_type (str): Record type (A, AAAA, etc.)
            ttl (int|None): TTL in seconds
            line (str|None): Line (not used by ClouDNS)
            extra (dict|None): Extra parameters

        Returns:
            bool: True if successful, False otherwise
        """
        # Use default TTL if not specified
        if ttl is None:
            ttl = self.DEFAULT_TTL

        # For @ (root) records, use empty string
        host = "" if subdomain == "@" else subdomain

        params = {"domain-name": zone_id, "record-type": record_type, "host": host, "record": value, "ttl": ttl}
        data = self._request("/dns/add-record.json", **params)

        if data and data.get("status") == "Success":
            self.logger.info("Record created successfully")
            return True

        self.logger.error("Failed to create record: %s", data)
        return False

    def _update_record(self, zone_id, old_record, value, record_type, ttl, line, extra):
        # type: (str, dict, str, str, int | None, str | None, dict | None) -> bool
        """
        更新DNS记录

        Update existing DNS record.
        https://www.cloudns.net/wiki/article/60/

        Args:
            zone_id (str): Zone ID (domain name)
            old_record (dict): Existing record data
            value (str): New record value (IP address)
            record_type (str): Record type (A, AAAA, etc.)
            ttl (int|None): TTL in seconds
            line (str|None): Line (not used by ClouDNS)
            extra (dict|None): Extra parameters

        Returns:
            bool: True if successful, False otherwise
        """
        # Use default TTL if not specified
        if ttl is None:
            ttl = self.DEFAULT_TTL

        record_id = old_record.get("id")
        if not record_id:
            self.logger.error("Record ID not found in old_record")
            return False

        # Get original host from record
        host = old_record.get("host", "")

        params = {"domain-name": zone_id, "record-id": record_id, "host": host, "record": value, "ttl": ttl}
        data = self._request("/dns/mod-record.json", **params)

        if data and data.get("status") == "Success":
            self.logger.info("Record updated successfully")
            return True

        self.logger.error("Failed to update record: %s", data)
        return False

File: rust/src/provider/cloudns.rs
use std::collections::BTreeMap;

use serde_json::Value;

use crate::error::{Error, Result};
use crate::http::{Method, form_encode};

use super::base::{CrudProvider, ProviderContext, RecordRequest, value_to_string};

pub struct CloudnsProvider<'a> {
    context: ProviderContext<'a>,
    zones: BTreeMap<String, String>,
}

impl<'a> CloudnsProvider<'a> {
    pub fn new(context: ProviderContext<'a>) -> Result<Self> {
        if context.id.is_empty() || context.token.is_empty() {
            return Err(Error::Config(
                "ClouDNS auth id and password must be configured".to_owned(),
            ));
        }
        Ok(Self {
            context,
            zones: BTreeMap::new(),
        })
    }

    fn api(&self, path: &str, mut parameters: BTreeMap<String, String>) -> Result<Option<Value>> {
        parameters.insert("auth-id".to_owned(), self.context.id.to_owned());
        parameters.insert("auth-password".to_owned(), self.context.token.to_owned());
        let response = self.context.send_json(
            Method::Post,
            path,
            &BTreeMap::new(),
            Some(form_encode(&parameters)),
            BTreeMap::from([(
                "content-type".to_owned(),
                "application/x-www-form-urlencoded".to_owned(),
            )]),
        )?;
        if response.get("status").and_then(Value::as_str) == Some("Failed") {
            let description = response
                .get("statusDescription")
                .and_then(Value::as_str)
                .unwrap_or("unknown error");
            // This exact lookup miss is documented at cloudns.net/wiki/article/134/.
            if path == "/dns/get-zone-info.json" && description == "Missing domain-name" {
                return Ok(None);
            }
            return Err(Error::Provider(format!(
                "ClouDNS API error: {}",
                self.context.logger.mask(description)
            )));
        }
        Ok(Some(response))
    }
}

impl CrudProvider for CloudnsProvider<'_> {
    fn context(&self) -> &ProviderContext<'_> {
        &self.context
    }

    fn zone_cache(&mut self) -> &mut BTreeMap<String, String> {
        &mut self.zones
    }

    fn query_zone_id(&mut self, domain: &str) -> Result<Option<String>> {
        self.api(
            "/dns/get-zone-info.json",
            BTreeMap::from([("domain-name".to_owned(), domain.to_owned())]),
        )?
        .map(|response| {
            if response.get("name").and_then(Value::as_str) == Some(domain) {
                Ok(domain.to_owned())
            } else {
                Err(Error::Provider(
                    "ClouDNS returned invalid zone information".to_owned(),
                ))
            }
        })
        .transpose()
    }

    fn query_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        _main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<Option<Value>> {
        let host = if subdomain == "@" { "" } else { subdomain };
        let response = self
            .api(
                "/dns/records.json",
                BTreeMap::from([
                    ("domain-name".to_owned(), zone_id.to_owned()),
                    ("host".to_owned(), host.to_owned()),
                    ("type".to_owned(), request.record_type.to_owned()),
                ]),
            )?
            .ok_or_else(|| Error::Provider("ClouDNS returned no record list".to_owned()))?;
        // The API also uses [] for an empty record collection.
        if response.as_array().is_some_and(Vec::is_empty) {
            return Ok(None);
        }
        let records = response
            .as_object()
            .ok_or_else(|| Error::Provider("ClouDNS returned an invalid record list".to_owned()))?;
        for record in records.values() {
            let record_host = record
                .get("host")
                .and_then(Value::as_str)
                .ok_or_else(|| Error::Provider("ClouDNS record has no host".to_owned()))?;
            let record_type = record
                .get("type")
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
                .ok_or_else(|| Error::Provider("ClouDNS record has no type".to_owned()))?;
            let matches_host =
                record_host == host || (subdomain == "@" && matches!(record_host, "" | "@"));
            if matches_host && record_type == request.record_type {
                return Ok(Some(record.clone()));
            }
        }
        Ok(None)
    }

    fn create_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        _main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let response = self.api(
            "/dns/add-record.json",
            BTreeMap::from([
                ("domain-name".to_owned(), zone_id.to_owned()),
                ("record-type".to_owned(), request.record_type.to_owned()),
                (
                    "host".to_owned(),
                    if subdomain == "@" { "" } else { subdomain }.to_owned(),
                ),
                ("record".to_owned(), request.address.to_owned()),
                ("ttl".to_owned(), request.ttl.unwrap_or(60).to_string()),
            ]),
        )?;
        if response
            .as_ref()
            .and_then(|response| response.get("status"))
            .and_then(Value::as_str)
            == Some("Success")
        {
            Ok(())
        } else {
            Err(Error::Provider(
                "ClouDNS failed to create record".to_owned(),
            ))
        }
    }

    fn update_record(
        &mut self,
        zone_id: &str,
        record: &Value,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let record_id = record
            .get("id")
            .and_then(value_to_string)
            .ok_or_else(|| Error::Provider("ClouDNS record has no id".to_owned()))?;
        let response = self.api(
            "/dns/mod-record.json",
            BTreeMap::from([
                ("domain-name".to_owned(), zone_id.to_owned()),
                ("record-id".to_owned(), record_id),
                (
                    "host".to_owned(),
                    record
                        .get("host")
                        .and_then(value_to_string)
                        .unwrap_or_default(),
                ),
                ("record".to_owned(), request.address.to_owned()),
                ("ttl".to_owned(), request.ttl.unwrap_or(60).to_string()),
            ]),
        )?;
        if response
            .as_ref()
            .and_then(|response| response.get("status"))
            .and_then(Value::as_str)
            == Some("Success")
        {
            Ok(())
        } else {
            Err(Error::Provider(
                "ClouDNS failed to update record".to_owned(),
            ))
        }
    }
}
